Obsidian/Net/MinecraftStream.Writing.cs

```diff
@@ -35,38 +36,96 @@ public void WriteByte(Enum value)
-        var varLength = packet.Id.GetVarIntLength();
-
-        using var tempStream = new MinecraftStream();
-        packet.Serialize(tempStream);
+        if (packet is BundledPacket bp)
+        {
+            // Wrap the bundle with delimiter packets and writes all content packets
+            this.WritePacket(new BundleDelimiterPacket());
+            foreach (var p in bp.Packets)
+                this.WritePacket(p);
+            this.WritePacket(new BundleDelimiterPacket());
+            return;
+        }
 
-        var length = isBundled ? varLength : varLength + (int)tempStream.Length;
+        using MinecraftStream dataStream = new();
+        dataStream.WriteVarInt(packet.Id);
+        packet.Serialize(dataStream);
+        int length = (int)dataStream.Length;
 
         this.Lock.Wait();
-        this.WriteVarInt(length);
-        this.WriteVarInt(packet.Id);
 
-        tempStream.Position = 0;
-        tempStream.CopyTo(this);
-
-        if (isBundled)
-        {
-            this.WriteVarInt(length);
-            this.WriteVarInt(packet.Id);
-        }
+        this.WriteVarInt(length);
+        dataStream.Position = 0;
+        dataStream.CopyTo(this);
 
         this.Lock.Release();
     }
 
     public void WriteAttributeModifier(AttributeModifier attributeModifier) =>
         AttributeModifier.Write(attributeModifier, this);
 
+    /// <summary>
+    /// Writes a full packet to the stream with compression enabled. Note that this method is NOT
+    /// equivalent to <see cref="WritePacket"/> even if the packet size is less than <paramref name="compressionThreshold"/>.
+    /// See <see href="https://minecraft.wiki/w/Minecraft_Wiki:Projects/wiki.vg_merge/Protocol#With_compression"/>.
+    /// </summary>
+    /// <param name="packet">The packet to write.</param>
+    /// <param name="compressionThreshold">The threshold of packet size at which to compress the packet.</param>
     public void WriteCompressedPacket(IClientboundPacket packet, int compressionThreshold)
     {
-        throw new NotImplementedException();
+        if (packet is BundledPacket bp)
+        {
+            // Wrap the bundle with delimiter packets and writes all content packets
+            this.WriteCompressedPacket(new BundleDelimiterPacket(), compressionThreshold);
+            foreach (var p in bp.Packets)
+                this.WriteCompressedPacket(p, compressionThreshold);
+            this.WriteCompressedPacket(new BundleDelimiterPacket(), compressionThreshold);
+            return;
+        }
+
+        using MinecraftStream dataStream = new();
+        dataStream.WriteVarInt(packet.Id);
+        packet.Serialize(dataStream);
+        int dataLength = (int)dataStream.Length;
+
+        if (dataLength >= compressionThreshold)
+        {   // Compress the packet
+            using MinecraftStream compressedStream = new();
+            using (ZLibStream zlibStream = new(compressedStream, CompressionLevel.Optimal))
```

Comment:
One more thing, sorry 😅 Gotta pass `leaveOpen: true` to zlibstream so it doesn't close the backing stream as well